sympy/solvers/ode/riccati.py

```diff
@@ -0,0 +1,502 @@
+    return m, ybar, numy, deny, True
+
+
+def solve_aux_eq(numa, dena, numy, deny, x, m):
+    # Assume that the solution is of the type
+    # p(x) = C0 + C1*x + ... + Cm*x**m
+    psyms = symbols(f'C0:{m}', cls=Dummy)
+    domain = ZZ[psyms]
+    psyms += (1, )
+    psol = 0
+    for i in range(m + 1):
+        psol += Poly(psyms[i]*x**i, x, domain=domain)
+
+    # Eq (5.16) in Thesis - Pg 81
+    auxeq = (dena*(numy.diff(x)*deny - numy*deny.diff(x) + numy**2) - numa*deny**2)*psol
+    if m >= 1:
+        px = psol.diff(x)
+        auxeq += 2*px*numy*deny*dena
+    if m >= 2:
+        auxeq += px.diff(x)*deny**2*dena
+    if m != 0:
+        # m is a non-zero integer. Find the constant terms using undetermined coefficients
+        return psol, linsolve_dict(auxeq.all_coeffs(), psyms, x), True
+    else:
+        # m == 0 . Check if 1 (x**0) is a solution to the auxiliary equation
+        return S(1), auxeq, auxeq == 0
+
+
+def solve_riccati(fx, x, b0, b1, b2):
+    # Step 1 : Convert to Normal Form
+    a = -b0*b2 + b1**2/4 - b1.diff(x)/2 + 3*b2.diff(x)**2/(4*b2**2) + b1*b2.diff(x)/(2*b2) - b2.diff(x, 2)/(2*b2)
+    a_t = a.together()
+    num, den = [Poly(e, x, extension=True) for e in a_t.as_numer_denom()]
+    num, den = num.cancel(den, include=True)
+
+    # Step 2
+    presol = []
+
+    # Step 3 : a(x) is 0
+    if num == 0:
+        presol.append(1/(x + Dummy('C1')))
+
+    # Step 4 : a(x) is a non-zero constant
+    elif x not in num.free_symbols.union(den.free_symbols):
+        presol.extend([sqrt(a), -sqrt(a)])
+
+    # Step 5 : Find poles and valuation at infinity
+    poles = find_poles(num, den, x)
+    inf_mul = poles.get(oo, 0)
+    poles, muls = list(poles.keys()), list(poles.values())
+    val_inf = val_at_inf(num, den, x)
+
+    if len(poles):
+        # Check necessary conditions (outlined in the module docstring)
+        if check_necessary_conds(val_inf, muls):
+            raise ValueError("Rational Solution doesn't exist")
+
+        # Step 6
+        # Construct c-vectors for each singular point
+        c = construct_c(num, den, x, poles, muls)
+
+        # Construct d vectors for each singular point
+        d = construct_d(num, den, x, val_inf, inf_mul)
+
+        # Step 7 : Iterate over all possible combinations and return solutions
+        for it in range(2**(len(poles) + 1)):
+            # For each possible combination, generate an array of 0's and 1's
+            # where 0 means pick 1st choice and 1 means pick the second choice.
+            choice = [int(x) for x in bin(it)[2:].zfill(len(poles) + 1)]
+
+            # Step 8 and 9 : Compute m and ybar
+            m, ybar, numy, deny, exists = compute_degree(x, poles, choice, c, d, -val_inf//2)
+
+            # Step 10 : Check if a valid solution exists. If yes, also check
+            # if m is a non-negative integer
+            if exists and S(m).is_nonnegative == True and S(m).is_integer == True:
```

Comment:
In some cases `m` here can be an algebraic expression that could be simplified to zero or perhaps another integer e.g.:
```python
In [6]: m
Out[6]: 
             _________________________________           _________________________________
            ╱     7348059520   716670080⋅√105           ╱     716670080⋅√105   7348059520 
           ╱      ────────── - ──────────────          ╱      ────────────── + ────────── 
          ╱           9              9                ╱             9              9      
         ╱    1 + ───────────────────────────        ╱    1 + ─────────────────────────── 
        ╱          265995800   25937800⋅√105        ╱          25937800⋅√105   265995800  
       ╱           ───────── - ─────────────       ╱           ───────────── + ─────────  
     ╲╱                3             3           ╲╱                  3             3      
-3 + ───────────────────────────────────────── + ─────────────────────────────────────────
                         2                                           2                    

In [7]: m.is_integer
Out[7]: True

In [8]: m.is_zero
Out[8]: True

In [9]: m.n()
Out[9]: 0.e-125

In [10]: minpoly(m)
Out[10]: x
```
The `is_integer` check gives true here because `is_zero` gives True and that is because `is_zero` checks `minpoly`. Subsequent code fails due to:
```python
  File "/Users/enojb/current/sympy/sympy/my/ricattic.py", line 137, in main
    sol = solve_riccati(yf, x, q0, q1, q2)
  File "/Users/enojb/current/sympy/sympy/sympy/solvers/ode/riccati.py", line 476, in solve_riccati
    psol, coeffs, exists = solve_aux_eq(num, den, numy, deny, x, m)
  File "/Users/enojb/current/sympy/sympy/sympy/solvers/ode/riccati.py", line 404, in solve_aux_eq
    psyms = symbols(f'C0:{m}', cls=Dummy)
  File "/Users/enojb/current/sympy/sympy/sympy/core/symbol.py", line 728, in symbols
    a = 0 if not a else int(a)
```
This only works for `m` because it is equal to zero. An algebraic expression equal to another integer would not work:
```python
In [11]: m2 = 2 + m

In [12]: m2
Out[12]: 
             _________________________________           _________________________________
            ╱     7348059520   716670080⋅√105           ╱     716670080⋅√105   7348059520 
           ╱      ────────── - ──────────────          ╱      ────────────── + ────────── 
          ╱           9              9                ╱             9              9      
         ╱    1 + ───────────────────────────        ╱    1 + ─────────────────────────── 
        ╱          265995800   25937800⋅√105        ╱          25937800⋅√105   265995800  
       ╱           ───────── - ─────────────       ╱           ───────────── + ─────────  
     ╲╱                3             3           ╲╱                  3             3      
-1 + ───────────────────────────────────────── + ─────────────────────────────────────────
                         2                                           2                    

In [13]: m2.is_integer

In [14]: m2.n()
Out[14]: 2.00000000000000

In [15]: minpoly(m2)
Out[15]: x - 2
```

Reply:
So what should I use? `.n()` with a threshold?

Reply:
Poly with extension=True will automatically simplify any algebraic expression:
```python
In [6]: (1 + sqrt(2))**2 + (1 - sqrt(2))**2
Out[6]: 
        2           2
(1 - √2)  + (1 + √2) 

In [7]: Poly(_, x, extension=True)
Out[7]: Poly(6, x, domain='ZZ')
```
If `m` had been computed using `Poly` then it would already be simplified to a rational if possible.

Reply:
This problem has been solved now using `Poly`.